File: R/worker.R
#' Initialize a jiebaRS worker
#'
#' This function can initialize a jiebaRS worker. See Details for more information.
#'
#' @details
#' **The `qmax` argument is not supported.** Although `jiebaR` documented
#' `qmax` for query workers, the value was never actually passed to the
#' underlying segmentation call. Similarly, the `jieba-rs` backend implements
#' search-mode segmentation without a configurable query threshold. To avoid
#' user confusion, jiebaRS omits the qmax argument entirely rather than
#' retaining a no-op parameter.
#'
#' `jieba-rs` does not expose dedicated public implementations for `mp` or
#' `hmm` workers. `jiebaRS` therefore maps `mp` to `cut(..., false)` and `hmm`
#' to `cut(..., true)`. This is a compatibility approximation rather than a
#' byte-for-byte reimplementation of `jiebaR`, and `jiebaRS` warns once per R
#' session when either type is requested.
#'
#' `tag` workers use `jieba-rs` tagging on top of the default mixed
#' segmentation path, which is the closest public behavior to `jiebaR`.
#'
#' `stop_word` and `stop_word_file` can be both supplied at once and then
#' be merged together. Then they will be normalized.
#'
#' In `jiebaRS`, `hmm` accepts either a logical scalar or a file path. A
#' logical value controls whether the underlying `jieba-rs`
#' segmentation/tagging pipeline may fall back to HMM for unknown terms. A
#' character scalar is interpreted as a path to a custom HMM model file and
#' enables HMM fallback with that model. The flag affects `mix` and `query`
#' workers directly, `tag` workers through the underlying mixed tagging path,
#' and `keywords` workers through TF-IDF keyword extraction. `mp`, `hmm`, and
#' `full` workers ignore the runtime switch because their `jieba-rs` backends
#' do not use this runtime switch.
#'
#' `dict` and `user` load dictionary files at worker creation time. `dict`
#' *replaces* the embedded main dictionary entirely; `user` accepts one or more
#' file paths whose entries are appended, in the supplied order, to whatever
#' main dictionary is in place (default or custom `dict`). Main dictionary
#' entries use `word [freq] [tag]`; an omitted frequency defaults to `1`. User
#' dictionary entries may use `word`, `word freq`, `word tag` (the legacy
#' two-column `jiebaR` format), or `word freq tag`. User frequencies must be
#' positive integers; when omitted, a frequency is inferred from the current
#' dictionary. Files must be UTF-8; a leading byte order mark (BOM) is allowed.
#'
#' @param type Worker type. Supported values are `"mix"`, `"mp"`, `"hmm"`,
#'   `"full"`, `"query"`, `"tag"`, `"keywords"`, and `"textrank"`.
#'   Default is `"mix"`.
#' @param stop_word Optional character vector of stop words supplied directly.
#' @param stop_word_file Optional file path containing one stop word per line.
#' @param hmm Logical scalar or character scalar. If logical, controls whether
#'   to enable HMM fallback for unknown terms. If character, must be a path to a
#'   custom HMM model file compatible with `jieba-rs`'s `hmm.model` format, and
#'   HMM fallback is enabled with that model. Default is `TRUE`.
#' @param topn Integer. The number of terms returned by `keywords` and
#'   `textrank` workers. Default is `5`.
#' @param min_keyword_length Positive integer. The minimum number of Unicode
#'   scalar values in terms returned by `keywords` and `textrank` workers.
#'   Default is `2`.
#' @param idf Optional character scalar. A path to a custom IDF dictionary
#'   file for `keywords` workers. Each line should be `word idf_value`. When
#'   `NULL`, the embedded default IDF dictionary is used. Ignored by non-keyword
#'   workers. Default is `NULL`.
#' @param dict Optional character scalar. A path to a custom main dictionary
#'   file that *replaces* the embedded dictionary. Each line should be
#'   `word [freq] [tag]` (whitespace-separated; `freq` defaults to `1`, `tag`
#'   defaults to empty). When `NULL`, the embedded dictionary is used. Default
#'   is `NULL`.
#' @param user Optional character vector containing one or more paths to user
#'   dictionary files whose entries are *appended* to the main dictionary in
#'   the supplied order. Supported line formats are `word`, `word freq`,
#'   `word tag`, and `word freq tag`. Omitted frequencies are inferred from the
#'   current dictionary. Default is `NULL`.
#' @param symbol Logical. Whether to keep symbol-like tokens in the sentence. Default is `FALSE`.
#' @param bylines [Deprecated] compatibility argument retained from `jiebaR`.
#'   `jiebaRS` no longer uses this value; control batch aggregation directly
#'   in specific functions.
#'
#' @return A `jieba_worker` S3 object.
#' @export
worker <- function(
  type = c("mix", "mp", "hmm", "full", "query", "tag", "keywords", "textrank"),
  stop_word = NULL,
  stop_word_file = NULL,
  hmm = TRUE,
  topn = 5L,
  min_keyword_length = 2L,
  idf = NULL,
  dict = NULL,
  user = NULL,
  symbol = FALSE,
  bylines = FALSE
) {
  type <- rlang::arg_match(type)
  stop_words <- normalize_stop_words(stop_word, stop_word_file)
  hmm_model <- ""

  if (identical(type, "mp")) {
    cli::cli_warn(
      paste(
        "`worker(type = 'mp')` is currently mapped to `jieba-rs`",
        "`cut(..., false)` because `jieba-rs` does not expose a dedicated",
        "`mp` segmenter. Results may differ from `jiebaR`."
      ),
      .frequency = "once",
      .frequency_id = "worker_mp_mapping"
    )
  }

  if (identical(type, "hmm")) {
    cli::cli_warn(
      paste(
        "`worker(type = 'hmm')` is currently mapped to `jieba-rs`",
        "`cut(..., true)` because `jieba-rs` does not expose a dedicated",
        "`hmm` segmenter. Results may differ from `jiebaR`."
      ),
      .frequency = "once",
      .frequency_id = "worker_hmm_mapping"
    )
  }

  if (rlang::is_string(hmm)) {
    if (!file.exists(hmm)) {
      cli::cli_abort("`hmm` must point to an existing custom HMM model file.")
    }
    hmm_model <- enc2utf8(hmm)
    hmm <- TRUE
  } else if (!rlang::is_bool(hmm)) {
    cli::cli_abort("`hmm` must be `TRUE`, `FALSE`, or a custom HMM model file path.")
  }

  if (!rlang::is_integerish(topn, n = 1) || topn < 0) {
    cli::cli_abort("`topn` must be a non-negative integer.")
  }
  topn <- as.integer(topn)

  if (
    !rlang::is_integerish(min_keyword_length, n = 1) ||
      min_keyword_length < 1
  ) {
    cli::cli_abort("`min_keyword_length` must be a positive integer.")
  }
  min_keyword_length <- as.integer(min_keyword_length)

  idf_path <- ""
  if (!is.null(idf)) {
    if (!rlang::is_string(idf)) {
      cli::cli_abort("`idf` must be `NULL` or a path to an IDF dictionary file.")
    }
    if (!file.exists(idf)) {
      cli::cli_abort("`idf` must point to an existing IDF dictionary file.")
    }
    idf_path <- enc2utf8(idf)
  }

  dict_path <- ""
  if (!is.null(dict)) {
    if (!rlang::is_string(dict)) {
      cli::cli_abort("`dict` must be `NULL` or a path to a main dictionary file.")
    }
    if (!file.exists(dict)) {
      cli::cli_abort("`dict` must point to an existing main dictionary file.")
    }
    dict_path <- enc2utf8(dict)
  }

  user_paths <- character()
  if (!is.null(user)) {
    if (!rlang::is_character(user) || rlang::is_empty(user) || anyNA(user)) {
      cli::cli_abort(
        "`user` must be `NULL` or a non-empty character vector of user dictionary file paths."
      )
    }
    if (!all(file.exists(user))) {
      cli::cli_abort("Every path in `user` must point to an existing user dictionary file.")
    }
    user_paths <- enc2utf8(user)
  }

  if (!rlang::is_bool(symbol)) {
    cli::cli_abort("`symbol` must be `TRUE` or `FALSE`.")
  }

  if (!missing(bylines)) {
    cli::cli_warn(
      paste(
        "`bylines` is deprecated in `jiebaRS` and no longer has any effect.",
        "Control batch aggregation explicitly in specific functions."
      )
    )
  }

  classes <- c(
    switch(
      type,
      keywords = "jieba_keywords",
      textrank = "jieba_textrank",
      tag = "jieba_tagger",
      "jieba_segmenter"
    ),
    "jieba_worker"
  )

  structure(
    list(
      ptr = new_worker(
        type,
        hmm,
        hmm_model,
        idf_path,
        dict_path,
        user_paths,
        topn,
        min_keyword_length,
        stop_words
      ),
      type = type,
      config = list(
        hmm = hmm,
        hmm_model = if (nzchar(hmm_model)) hmm_model else NULL,
        topn = topn,
        min_keyword_length = min_keyword_length,
        idf = if (nzchar(idf_path)) idf_path else NULL,
        dict = if (nzchar(dict_path)) dict_path else NULL,
        user = if (length(user_paths)) user_paths else NULL,
        stop_word = stop_words,
        symbol = symbol,
        bylines = bylines
      )
    ),
    class = classes
  )
}

File: src/rust/src/worker.rs
use std::io::Cursor;

use ahash::AHashSet;

use extendr_api::prelude::*;
use extendr_api::{Error, Result};
use jieba_rs::{HmmModel, Jieba, KeywordExtractConfig, TextRank, TfIdf};

use crate::file_reader::{read_dictionary, read_idf_dictionary, read_utf8_file, DictionaryKind};

pub const WORKER_ABI_VERSION: i32 = 1;

/// Construction parameters for [`JiebaWorker`].
pub struct WorkerConfig<'a> {
    pub worker_type: &'a str,
    pub use_hmm: bool,
    pub hmm_model: &'a str,
    pub idf_path: &'a str,
    pub dict_path: &'a str,
    pub user_paths: Vec<String>,
    pub top_n: u32,
    pub min_keyword_length: u32,
    pub stop_words: Vec<String>,
}

#[derive(Clone, Copy)]
pub enum WorkerFamily {
    Segment(SegmentMode),
    Tag,
    Keywords,
    TextRank,
}

#[derive(Clone, Copy)]
pub enum SegmentMode {
    Mix,
    Mp,
    Hmm,
    Full,
    Query,
}

impl WorkerFamily {
    fn from_type(worker_type: &str) -> Result<Self> {
        match worker_type {
            "mix" => Ok(Self::Segment(SegmentMode::Mix)),
            "mp" => Ok(Self::Segment(SegmentMode::Mp)),
            "hmm" => Ok(Self::Segment(SegmentMode::Hmm)),
            "full" => Ok(Self::Segment(SegmentMode::Full)),
            "query" => Ok(Self::Segment(SegmentMode::Query)),
            "tag" => Ok(Self::Tag),
            "keywords" => Ok(Self::Keywords),
            "textrank" => Ok(Self::TextRank),
            _ => Err(Error::Other(format!(
                "Unsupported worker type `{worker_type}`. Supported types are `mix`, `mp`, `hmm`, `full`, `query`, `tag`, `keywords`, and `textrank`."
            ))),
        }
    }
}

#[extendr]
pub struct JiebaWorker {
    pub engine: Jieba,
    pub family: WorkerFamily,
    pub use_hmm: bool,
    pub top_n: usize,
    pub min_keyword_length: usize,
    pub stop_words: AHashSet<String>,
    pub keyword_extractor: Option<TfIdf>,
    pub textrank_extractor: Option<TextRank>,
    version: i32,
}

impl JiebaWorker {
    pub fn new(config: WorkerConfig<'_>) -> Result<Self> {
        let WorkerConfig {
            worker_type,
            use_hmm,
            hmm_model,
            idf_path,
            dict_path,
            user_paths,
            top_n,
            min_keyword_length,
            stop_words,
        } = config;

        let family = WorkerFamily::from_type(worker_type)?;
        let top_n = top_n as usize;
        let min_keyword_length = min_keyword_length as usize;
        let stop_words: AHashSet<String> = stop_words
            .into_iter()
            .map(|word| word.to_lowercase())
            .collect();
        let keyword_stop_words = stop_words.iter().cloned().collect();
        let textrank_stop_words = stop_words.iter().cloned().collect();

        let keyword_extractor = match family {
            WorkerFamily::Keywords => {
                let config = KeywordExtractConfig::builder()
                    .set_stop_words(keyword_stop_words)
                    .use_hmm(use_hmm)
                    .min_keyword_length(min_keyword_length)
                    .build();

                let extractor = if idf_path.is_empty() {
                    let mut e = TfIdf::default();
                    *e.config_mut() = config;
                    e
                } else {
                    let contents = read_idf_dictionary(idf_path)?;
                    let mut reader = Cursor::new(contents);
                    TfIdf::new(Some(&mut reader), config)
                };
                Some(extractor)
            }
            WorkerFamily::Segment(_) | WorkerFamily::Tag | WorkerFamily::TextRank => None,
        };

        let textrank_extractor = match family {
            WorkerFamily::TextRank => {
                let config = KeywordExtractConfig::builder()
                    .set_stop_words(textrank_stop_words)
                    .use_hmm(use_hmm)
                    .min_keyword_length(min_keyword_length)
                    .build();
                Some(TextRank::new(5, config))
            }
            WorkerFamily::Segment(_) | WorkerFamily::Tag | WorkerFamily::Keywords => None,
        };

        let mut engine = if dict_path.is_empty() {
            // Default: use the embedded dictionary.
            Jieba::new()
        } else {
            // `dict` replaces the main dictionary entirely.
            let entries = read_dictionary(dict_path, DictionaryKind::Main)?;
            let normalized = entries
                .into_iter()
                .map(|entry| {
                    let frequency = entry.frequency.unwrap_or(1);
                    match entry.tag {
                        Some(tag) => format!("{} {frequency} {tag}\n", entry.word),
                        None => format!("{} {frequency}\n", entry.word),
                    }
                })
                .collect::<String>();
            let mut reader = Cursor::new(normalized);
            Jieba::with_dict(&mut reader).map_err(|err| {
                Error::Other(format!(
                    "Failed to load custom main dictionary `{dict_path}`: {err}"
                ))
            })?
        };

        // `user` appends to whatever dictionary is in place (default or
        // custom `dict`).
        for user_path in &user_paths {
            for entry in read_dictionary(user_path, DictionaryKind::User)? {
                engine.add_word(&entry.word, entry.frequency, entry.tag.as_deref());
            }
        }

        if !hmm_model.is_empty() {
            let contents = read_utf8_file(hmm_model, "custom HMM model")?;
            let mut reader = Cursor::new(contents);
            let model = HmmModel::load(&mut reader).map_err(|err| {
                Error::Other(format!(
                    "Failed to load custom HMM model `{hmm_model}`: {err}"
                ))
            })?;
            engine.set_hmm_model(model);
        }

        Ok(Self {
            engine,
            family,
            use_hmm,
            top_n,
            min_keyword_length,
            stop_words,
            keyword_extractor,
            textrank_extractor,
            version: WORKER_ABI_VERSION,
        })
    }

    pub fn validate(&self) -> Result<()> {
        if self.version != WORKER_ABI_VERSION {
            return Err(Error::Other(
                "Worker ABI version mismatch. Please create a new worker.".to_string(),
            ));
        }

        Ok(())
    }

    pub fn keep_token(&self, token: &str) -> bool {
        token != " " && !self.stop_words.contains(token)
    }
}
